### ISSUE TITLE
Keep relay logs, when MasterFailoverDetachReplicaMasterHost: true

### ISSUE THREAD

**Issue by ghost** [openark/orchestrator#736]
When **MasterFailoverDetachReplicaMasterHost: true**, the **stop slave** and **change master** are sent to promoted server and all relay logs are removed.
It also happens even if, **ApplyMySQLPromotionAfterMasterFailover": false**.
From mysql docs:

> Prior to MySQL 5.7.4, CHANGE MASTER TO deletes all relay log files and starts a new one, unless you specify RELAY_LOG_FILE or RELAY_LOG_POS. In that case, relay log files are kept; the relay_log_purge global variable is set silently to 0. In MySQL 5.7.4 and later, relay logs are preserved if at least one of the slave SQL thread and the slave I/O thread is running; if both threads are stopped, all relay log files are deleted unless at least one of RELAY_LOG_FILE or RELAY_LOG_POS is specified.

Does it make sense to add "RELAY_LOG_FILE=current_relay_log_file", which will do nothing, but keep the relay logs after **change master**?

Also in case of GTID, MASTER_AUTO_POSITION=0 should be added otherwise:

> ERROR 1776 (HY000): Parameters MASTER_LOG_FILE, MASTER_LOG_POS, RELAY_LOG_FILE and RELAY_LOG_POS cannot be set when MASTER_AUTO_POSITION is active.

The full statement may look like this:
`change master to master_host='\\dbserver', master_port=3306, relay_log_file='relay-bin.000002', master_auto_position=0;`

**Comment by shlomi-noach** [comment]
@pasha167 there's a lot of options here. Can you please backtrack and explain what you are trying to achieve and what is the problem you see?

**Comment by ghost** [comment]
@shlomi-noach, my goal is to let promoted server to apply all trx it may have in relay log.
But I also want to be sure, that crashed master is detached. So I set the following parameters:
```
  "FailMasterPromotionIfSQLThreadNotUpToDate": true,
  "ApplyMySQLPromotionAfterMasterFailover": false,
  "MasterFailoverDetachReplicaMasterHost": true,
```
After crash I've got the following logs:

> 2018-11-23 08:10:19 INFO topology_recovery: RecoverDeadCoMaster: mustPromoteOtherCoMaster? false
> 2018-11-23 08:10:19 INFO topology_recovery: checking if should replace promoted replica with a better candidate
> 2018-11-23 08:10:19 INFO topology_recovery: + checking if promoted replica is the ideal candidate
> 2018-11-23 08:10:19 INFO topology_recovery: promoted replica Slave1 is the ideal candidate
> 2018-11-23 08:10:19 INFO topology_recovery: promoted instance Slave1 requires no further action
> 2018-11-23 08:10:19 INFO Will detach master host on Slave1. Detached key is //Master1
> 2018-11-23 08:10:19 INFO Stopped slave on Slave1, Self:mysql-bin.000001:15853335, Exec:mysql-bin.000001:62202446
> 2018-11-23 08:10:19 DEBUG ChangeMasterTo: will attempt changing master on Slave1 to //Master1, mysql-bin.000001:62202446
> 2018-11-23 08:10:19 INFO ChangeMasterTo: Changed master on Slave1 to: //Master1, mysql-bin.000001:62202446. GTID: true
> 2018-11-23 08:10:20 ERROR Promoted replica Slave1: sql thread is not up to date (relay logs still unapplied). Aborting promotion
> 2018-11-23 08:10:20 ERROR 2018-11-23 08:10:20 ERROR Promoted replica Slave1: sql thread is not up to date (relay logs still unapplied). Aborting promotion
> 
During detach all relay logs are lost due to **change master** statement. 

"MasterFailoverDetachReplicaMasterHost": true, does three things actually: stop slave, detach and clean relay logs. 
So I am looking for possibilities to have only "detach" and keep relay logs with running sql thread.

**Comment by shlomi-noach** [comment]
> So I am looking for possibilities to have only "detach" and keep relay logs with running sql thread.

Got it.

I happen to know of a contributor who has made a patch they intend to push upstream: upon failover, the newly promoted master will first consume all of the relay logs, before making the final promotion (setting read-only=0 etc.)

So I will wait for their contribution (which I hope will surface in the next few weeks).

**Comment by ghost** [comment]
Ok. That might solve my problem as well. Let's wait. Thank you.

**Comment by ghost** [comment]
@shlomi-noach, btw, as a proposition: does it make sense to add more details to the description of **MasterFailoverDetachReplicaMasterHost** in code and documentation? 
For me it was not expected that it cleans the relay logs. So some people may be confused as well.
That might be also better to decide after the patch. I just leave the comment here to not forget :).

**Comment by shlomi-noach** [comment]
> does it make sense to add more details 

It does indeed. Would you consider making a documentation PR? If not, I can do so later.

**Comment by shlomi-noach** [comment]
Regardless, let me consider your suggestion for setting `AUTO_POSITION=0` and `RELAY_LOG_FILE`. There is risk in setting `MASTER_AUTO_POSITION=0` because we lose the knowledge that a server used to use `AUTO_POSITION` and when reattaching we don't know whether we should keep it with `0` or change to `1`. 
The information can be encoded in the "detached" hostname.

**Comment by ghost** [comment]
Yes, I see the problem here.
After 5.7.4 the `change master` can be executed without stopping slave. With running `sql_thread` relay logs are not removed. So there is no need to add `RELAY_LOG_FILE` and change `AUTO_POSITION`:

> MySQL 5.7.4 and later, relay logs are preserved if at least one of the slave SQL thread and the slave I/O thread is running; if both threads are stopped, all relay log files are deleted unless at least one of RELAY_LOG_FILE or RELAY_LOG_POS is specified.

But for 5.6, if I am not wrong,`stop slave` is a must before `change master`.
Is it good idea to check version and act differently depending on it?? Also it means that for 5.6 the logs will still be removed.

PS: I would be glad to do PR and adjust documentation.

**Comment by shlomi-noach** [comment]
Since the replication will have been broken (this is a master failure scenario), IO thread will have been stopped. I wouldn't count on SQL thread to keep on running as the sole thing to keep relay logs from being purged on `change master to`.

**Comment by ghost** [comment]
Yes, you are right. 
I agree, none of the proposed solutions are perfect.
Regarding the losing value of auto_position...hard to estimate, how bad it is. Some people would accept it. Maybe add other parameter like MasterFailoverDetachReplicaMasterHostKeepRelayLogs that will keep the logs, but lose the auto_position?
Sorry, don't see good solution for now.

**Comment by MOON-CLJ** [comment]
@shlomi-noach we just found StopSlaveNicely do wait stuff for all slaves in sortedReplicasDataCenterHint。is it just done "will first consume all of the relay logs, before making the final promotion "?  cc @yangeagle

**Comment by MOON-CLJ** [comment]
ping @shlomi-noach

**Comment by shlomi-noach** [comment]
@MOON-CLJ sorry for the late response.

> is it just done "will first consume all of the relay logs, before making the final promotion "

I'm not sure I understand the question, and I think there's a language barrier and a translation problem here.
If I interpret the question correctly, then the answer is "yes".

**Comment by MOON-CLJ** [comment]
@shlomi-noach sorry, let me try again to explain my question。

1, we propose to add a wait logic in https://github.com/github/orchestrator/issues/757#issue-390603100 。then you reply "I've had multiple similar requests to have the same wait, but just before finalizing master promotion. " in https://github.com/github/orchestrator/issues/757#issuecomment-446923685 。and "I happen to know of a contributor who has made a patch they intend to push upstream: upon failover, the newly promoted master will first consume all of the relay logs, before making the final promotion " in https://github.com/github/orchestrator/issues/736#issuecomment-441419105 。
2, after some time, we found StopSlaveNicely https://github.com/github/orchestrator/blob/a2315bb61d5e96eb7c57d2ede2fa9130df10689c/go/inst/instance_topology_dao.go#L281 seems already done the wait logic。
3, so i want to know if we understand this correctly ? are we missing anything? 
what's the difference between the patch (what we wait for) and StopSlaveNicely?

**Comment by shlomi-noach** [comment]
> what's the difference between the patch (what we wait for) and StopSlaveNicely?

I haven't seen the patch yet. @akshaysuryawanshi do you know if this is expected to arrive any time soon?

@MOON-CLJ basically I agree that this would be a `StopSlaveNicely()` operation.
One thing to note is that `StopSlaveNicely()` can sometimes not work as expected. More correctly, `start slave until RELAY_LOG_FILE=... RELAY_LOG_POS=...` is not guaranteed to succeed. 

The following is my observation, though I'm not sure if it's the correct analysis: if the last retrieved relay log entry consists of a `table_map` (or something that begins a transaction), then `start slave until` can hang forever.
I found about that while automating some internal experiment. So I'd like to get more visibility into this sort of problem.

**Comment by akshaysuryawanshi** [comment]
@shlomi-noach, @jordanwheeler should be submitting a PR soon on this.
